enhancements/cli_manager.md

```diff
@@ -0,0 +1,144 @@
+---
+title: cli-manager
+authors:
+  - "@sallyom"
+reviewers:
+  - TBD
+  - "@soltysh"
+approvers:
+  - TBD
+  - "@soltysh"
+creation-date: 2019-12-03
+last-updated: 2019-12-03
+status: provisional
+---
+
+# cli-manager
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal is for a component to manage various tools available for OpenShift and related services.
+A CLI manager is desired so that a user can discover, install and upgrade tools easily and from a single
+location.  Each component is responsible for serving its files and registering information regarding supplied binaries
+to a central index location.  The CLI manager will retrieve all the binaries from the index location.
+Additionally, CLI manager must be supported in disconnected environments.  Currently, we provide disconnected environments 
+cli-artifacts and oc download links that do not require anything outside the cluster.  The goal is to provide the same for other
+binaries/tools/CLIs.  The CLI manager will provide a mechanism to provide, list, install, and upgrade OpenShift tools/plugins/CLIs from
+within a cluster.  The [ConsoleCLIDownload CRD](https://github.com/openshift/api/blob/master/console/v1/0000_10_consoleclidownload.crd.yaml) 
+may be extended to meet the requirements of a new CLI manager.
+
+## Motivation
+
+As more services are created on top of OpenShift, more CLIs are introduced to simplify interaction with these services.
+Some current examples are oc, kubectl, odo, istio, tekton, and knative.  It is difficult for users to discover what tools exist, 
+where to download them from and which version they should download.  We need to simplify as much as possible the interaction
+of services on OpenShift.  We need a mechanism for providing and consuming tools that is simple to add on to as new CLIs are 
+developed from a variety of sources - and this needs to be managed from within a cluster.
+This is the motivation for the CLI manager.
+
+### Goals
+
+Each component wishing to provide customers with their binaries will build and publish them via an official channel to an
+index location from which the CLI manager will obtain the files.
+
+Two routes for index:    
+1.  Central index that has links to where artifacts are stored
```

Comment:
If an index exists that has the list of CLIs and their location, what would be ConsoleCLIDownload used for?

Reply:
The CRD will hold info for each CLI (available versions, download urls, description, other) that can be gathered via `oc get climanageddownload okd -o yaml` for example.

Reply:
I'd imagine we'd deprecate `ConsoleCLIDownload` in favour of the central index. All the tools serving that information should be using the new central repository.

Reply:
Is there a case where an optional operator can not, or doesn't want to, ship their CLI through the central index?

Reply:
+1 for the central index which would remove the requirement from the user to have access and be logged into a cluster when she wants to list or install plugins.

Reply:
The cli-manager will be run within a cluster - so yes, users will have to be logged into a cluster to list or install CLI plugins.  This way users will know they are using the CLI version that is known good with whatever version their cluster is. It will be possible to install CLIs outside a cluster, but not through cli-manager.

Reply:
When I said `The cli-manager will be run within a cluster` that's not what I meant, really - each cluster will have its set of ClusterCLI CRs - those hold the the known-good CLIs (compatible or latest version, etc) for that cluster.  `oc` will access those when connected to a cluster.  Also, I've updated this proposal to move away from `cli-manager` since what we're actually proposing is extending/adding to `oc` something like `oc clustercli list|search|install`